### DISCUSSION TITLE
s_client does not output all data

### DISCUSSION THREAD

**Discussion by EnDe** [openssl/openssl#25305] [Q&A] [answered]
Hi Team,

when calling `echo|openssl s_client -connect localhost:443` we get a bunch of information about the handshake and the final session, for example Protocol: Server certificate, Cipher:, Session-ID:, PSK identity:, and so on.

When the same command is executed by another process, this information is sometimes missing.

After some debugging, I guess it has to do with the way openssl writes its data.
Can you confirm, that some data is written on STDOUT and others on a terminal device directly?

Here are examples to reproduce it:
* `echo|openssl s_client -connect localhost:443` 
* `perl -le 'system("echo|openssl s_client -connect localhost:443");'`  # same result as before
* `perl -le 'qx(echo | openssl s_client -connect localhost:443);'` # missing data

As I'm not used to the concepts behind openssl, I cannot qualify this issue. At least to me it is anoying as tools using openssl behave different when called in a terminal or without (for example cron).

Ciao

**Comment by t8m** [comment]
Some data is written to STDOUT, some information is written to STDERR.

I do not think we write to the terminal device directly.

**Comment by EnDe** [comment]
Thanks. The data I'm asking for is not on STDERR, definitely.
Simply test with `echo|openssl s_client -connect localhost:443>/dev/null`

**Reply by t8m** [reply]
Can you please give a concrete example?

**Comment by EnDe** [comment]
please see the example commands I posted in the question, use the last two examples with perl, and you see directly what I mean.
For the sake of whatever, a snippet what is missing:

```
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 46B184E1CD48F6B66EE0DF52F375A4D42FD174E7B07DE31C0FBB180CC7122DD1
    Session-ID-ctx: 
    Resumption PSK: 30704946D1F0552DEFD5FF453601CFAC3A5F081154DC0E64563A2C51F09B404B19EBF93DED8F6D76D1E7BF94AFA5C0B5
    PSK identity: None
    PSK identity hint: None
````

**Reply by t8m** [reply]
Now this makes things more clear. The Post-Handshake message is not read if there is no readable stdin which is probably the case with qx(). - the s_client terminates before it reads the post handshake message.

**Comment by EnDe** [comment]
hmm, this might be "a" reason.
But when you issue `echo|openssl s_client -connect localhost:443` with cron, you get the same behaviour as with Perl's qx(), so it's not a Perl issue (don't forget the full path if you test with cron ;-)
So the question then would be: why does s_client terminate if there is no STDIN.
However, omit the `echo|` part, and its the same behaviour beside the coffee brake ...


BTW, the missing text is printed before s_client waits on STDIN (as I understand it)

**Reply by t8m** [reply]
Hmm... you're right that something else is in play here.

**Comment by t8m** [comment, marked as answer]
This works for me:
```
perl -le 'my $out = qx((echo;sleep 1) | openssl s_client -connect localhost:443); print $out;'
```

**Reply by t8m** [reply]
So this indicates the problem is with the STDIN being closed too early in the qx command.

**Reply by EnDe** [reply]
perfect wokraround, thanks.
